app/main/views/suppliers.py

```diff
@@ -91,10 +94,81 @@ def supplier_details():
     except APIError as e:
         abort(e.status_code)
     supplier['contact'] = supplier['contactInformation'][0]
+    country_name = get_country_name_from_country_code(supplier['registrationCountry'])
 
     return render_template(
         "suppliers/details.html",
         supplier=supplier,
+        country_name=country_name,
+    ), 200
+
+
+@main.route('/registered-address/edit', methods=['GET', 'POST'])
+@login_required
+def edit_registered_address():
+    try:
+        supplier = data_api_client.get_supplier(
+            current_user.supplier_id
+        )['suppliers']
+    except APIError as e:
+        abort(e.status_code)
+    supplier['contact'] = supplier['contactInformation'][0]
+
+    registered_address_form = EditRegisteredAddressForm()
+    registered_country_form = EditRegisteredCountryForm()
+
+    if request.method == 'POST':
+        address_valid = registered_address_form.validate_on_submit()
+        country_valid = registered_country_form.validate_on_submit()
+
+        if (address_valid and country_valid):
+            try:
+                data_api_client.update_supplier(
+                    current_user.supplier_id,
+                    registered_country_form.data,
+                    current_user.email_address,
+                )
+
+                data_api_client.update_contact_information(
+                    current_user.supplier_id,
+                    supplier['contact']['id'],
+                    registered_address_form.data,
+                    current_user.email_address
+                )
+            except APIError as e:
+                return render_template(
+                    "suppliers/registered_address.html",
+                    supplier=supplier,
+                    countries=COUNTRY_TUPLE,
+                    registered_address_form=registered_address_form,
+                    registered_country_form=registered_country_form,
+                    error=e.message,
+                ), 400
+
+            return redirect(url_for(".supplier_details"))
+
+        return render_template(
+            "suppliers/registered_address.html",
+            supplier=supplier,
+            countries=COUNTRY_TUPLE,
+            registered_address_form=registered_address_form,
+            registered_country_form=registered_country_form,
+            error=None,
+        ), 400
+
+    registered_address_form.address1.data = supplier['contact'].get('address1')
+    registered_address_form.city.data = supplier['contact'].get('city')
+    registered_address_form.postcode.data = supplier['contact'].get('postcode')
+
+    registered_country_form.registrationCountry.data = supplier.get('registrationCountry')
+
+    return render_template(
```

Comment:
We could probably refactor this down a bit so that we only have a single render_template on this one template, but not sure if it would be cleaner or not. The two POST-based ones could be combined relatively easily though.

Reply:
How were you thinking of refactoring the posts?

Reply:
You could do something like this for the just POST section:

```python
    if request.method == 'POST':
        error = None
        address_valid = registered_address_form.validate_on_submit()
        country_valid = registered_country_form.validate_on_submit()

        if address_valid and country_valid:
            try:
                data_api_client.update_supplier(
                    current_user.supplier_id,
                    registered_country_form.data,
                    current_user.email_address,
                )

                data_api_client.update_contact_information(
                    current_user.supplier_id,
                    supplier['contact']['id'],
                    registered_address_form.data,
                    current_user.email_address
                )
                
            except APIError as e:
                error = e
                
            else:
                return redirect(url_for(".supplier_details"))

        return render_template(
            "suppliers/registered_address.html",
            supplier=supplier,
            countries=COUNTRY_TUPLE,
            registered_address_form=registered_address_form,
            registered_country_form=registered_country_form,
            error=error.message if error else None,
        ), 400
```


Or something like this for the full method (although I think the flow gets a bit confusing to follow here, so this maybe isn't a good idea):
```python
def edit_registered_address():
    try:
        supplier = data_api_client.get_supplier(
            current_user.supplier_id
        )['suppliers']
    except APIError as e:
        abort(e.status_code)
    supplier['contact'] = supplier['contactInformation'][0]

    error = None
    http_status = 200
    registered_address_form = EditRegisteredAddressForm()
    registered_country_form = EditRegisteredCountryForm()

    if request.method == 'POST':
        address_valid = registered_address_form.validate_on_submit()
        country_valid = registered_country_form.validate_on_submit()

        if address_valid and country_valid:
            try:
                data_api_client.update_supplier(
                    current_user.supplier_id,
                    registered_country_form.data,
                    current_user.email_address,
                )

                data_api_client.update_contact_information(
                    current_user.supplier_id,
                    supplier['contact']['id'],
                    registered_address_form.data,
                    current_user.email_address
                )

            except APIError as e:
                error = e

            else:
                return redirect(url_for(".supplier_details"))

        http_status = 400

    else:
        registered_address_form.address1.data = supplier['contact'].get('address1')
        registered_address_form.city.data = supplier['contact'].get('city')
        registered_address_form.postcode.data = supplier['contact'].get('postcode')

        registered_country_form.registrationCountry.data = supplier.get('registrationCountry')

    return render_template(
        "suppliers/registered_address.html",
        supplier=supplier,
        countries=COUNTRY_TUPLE,
        registered_address_form=registered_address_form,
        registered_country_form=registered_country_form,
        error=error.message if error else None,
    ), http_status
```

Reply:
Ah nice. I totally forgot that `else` was a thing.

Reply:
Looking at it, I don't think that the flow in the second option is that confusing.